File: experiments/001_wow_fishing/run_test_a.py
"""Compare bite policies with identical preparation, observations and input checks."""

import argparse
from datetime import datetime, timezone
import hashlib
import json
import os
from pathlib import Path
import subprocess
import time
import uuid

ROOT = Path(__file__).resolve().parents[2]
BINARY = Path('/tmp/jev-fishing-live')


def invoke(arguments, timeout):
    result = subprocess.run([str(BINARY), *arguments], cwd=ROOT, capture_output=True,
                            text=True, timeout=timeout)
    events = []
    for line in result.stdout.splitlines():
        try:
            value = json.loads(line)
        except json.JSONDecodeError:
            continue
        if isinstance(value, dict) and 'event' in value:
            events.append(value)
    return result, events


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('--background', action='store_true')
    parser.add_argument('--jev', action='store_true', help='Test B: provider-backed bite decisions; shared deterministic preparation')
    args = parser.parse_args()
    label = 'B' if args.jev else 'A'
    if args.jev:
        if not os.environ.get('TYPESAFE_API_KEY') and (ROOT / '.env').is_file():
            for line in (ROOT / '.env').read_text().splitlines():
                name, sep, value = line.removeprefix('export ').partition('=')
                if sep and name.strip() == 'TYPESAFE_API_KEY':
                    os.environ['TYPESAFE_API_KEY'] = value.strip().strip('\"\'')
        if not os.environ.get('TYPESAFE_API_KEY'):
            raise SystemExit('Set TYPESAFE_API_KEY locally before Test B')
        os.environ['JEV_CALL_LIMIT'] = '120'
    mode = ['--background'] if args.background else []
    policy_mode = mode + (['--jev'] if args.jev else [])
    run_id = datetime.now(timezone.utc).strftime('%Y%m%dT%H%M%SZ') + '_' + uuid.uuid4().hex[:6]
    folder = ROOT / 'runs' / '001_wow_fishing' / ('test_' + label.lower() + '_' + run_id)
    folder.mkdir(parents=True)
    record = {'protocol': 'anchored-pixel-bite-only-v1', 'comparison_scope': 'bite_policy_only',
              'policy': 'jev' if args.jev else 'rules', 'call_budget': 120 if args.jev else 0, 'status': 'running', 'mode': 'targeted' if args.background else 'foreground',
              'input_mode': 'targeted_without_activation' if args.background else 'foreground_without_activation',
              'game_foreground_observed': False, 'focus_observations': 0,
              'target_seconds': 300, 'pre_go_runs': 1, 'provider_calls': 0,
              'started_at_utc': datetime.now(timezone.utc).isoformat(), 'cycles': []}
    source = ROOT / 'experiments' / '001_wow_fishing'
    record['source_hashes'] = {
        str(p.relative_to(source)): hashlib.sha256(p.read_bytes()).hexdigest()
        for p in [source/'live.swift', source/'decision.swift', source/'self_tests.swift', source/'motion.swift', source/'motion-fixtures.png', source/'jev.swift', source/'loot.swift', source/'loot-close.png', source/'loot-layout.png', source/'build.sh', source/'page-two.png', source/'rod-icon.png', source/'split-bobber.png', source/'split-bobber-before.png', source/'bobber-no-red.png', source/'bobber-no-red-before.png', source/'run_test_a.py',
                  source/'probes/background-click/Adapter.swift',
                  source/'probes/background-click/NativeWindowServerPreparation.swift',
                  source/'probes/background-click/NativeBackgroundClickTransport.swift']}
    print(f'Test {label} {record["mode"]}; results: {folder}', flush=True)
    def account(events):
        record['provider_calls'] += sum(e['event'] == 'jev_request' for e in events)
        for event in events:
            if event['event'] == 'focus_observed':
                record['focus_observations'] += 1
                record['game_foreground_observed'] |= event['game_foreground']
            if event['event'] == 'jev_response':
                record.setdefault('request_seconds', []).append(event['request_seconds'])
                for name, count in event.get('response', {}).get('usage', {}).items():
                    record[name] = record.get(name, 0) + count
        if args.jev:
            os.environ['JEV_CALL_LIMIT'] = str(max(0, 120-record['provider_calls']))
    started = None
    try:
        prego, events = invoke(['--check', *mode], 120)
        account(events)
        (folder/'pre-go.log').write_text(prego.stdout + prego.stderr)
        if prego.returncode or not any(e['event'] == 'pre_go_pass' for e in events):
            record['status'] = 'pre_go_failed'
            return
        started = time.monotonic()
        print('Pre-go passed once; autonomous 300-second test started.', flush=True)
        consecutive_failures = 0
        # Finish an in-flight cast rather than kill it with a mouse button held.
        # This bounds normal completion to 300 seconds plus at most one 45-second cycle.
        while time.monotonic()-started < 300:
            number = len(record['cycles']) + 1
            result, events = invoke(['--execute', '--prepared', *policy_mode], 45)
            account(events)
            (folder/f'cycle-{number:02d}.log').write_text(result.stdout + result.stderr)
            collected = any(e['event'] == 'loot_collected' for e in events)
            outcome = 'loot_collected' if collected else (events[-1]['event'] if events else 'process_failed')
            cycle = {'number': number, 'outcome': outcome, 'exit_code': result.returncode,
                     'elapsed_seconds': round(time.monotonic()-started, 2),
                     'run_path': next((e['output'] for e in events if e['event'] == 'ready'), None),
                     'item': next((e['item'] for e in events if e['event'] == 'loot_collected'), None),
                     'labels_observed': next((e.get('labels_observed', []) for e in events if e['event'] == 'loot_collected'), []),
                     'loot_clicks': next((e.get('loot_clicks', 1) for e in events if e['event'] == 'loot_collected'), 0)}
            record['cycles'].append(cycle)
            print(json.dumps(cycle), flush=True)
            consecutive_failures = 0 if collected else consecutive_failures+1
            if consecutive_failures >= 3:
                record['status'] = 'stopped_after_three_consecutive_failures'
                break
            # A native safety stop is terminal. Never repair a changed view by recasting.
            if result.returncode or any(e['event'].startswith('stopped_') or e['event'] in [
                    'loot_item_unconfirmed', 'loot_not_cleared', 'loot_layout_unconfirmed',
                    'loot_batch_limit', 'retrieval_unverified'] for e in events):
                record['status'] = 'stopped_for_review'
                break
            time.sleep(0.5)
        else:
            record['status'] = 'completed' if any(c['outcome'] == 'loot_collected' for c in record['cycles']) else 'no_verified_catches'
    except KeyboardInterrupt:
        record['status'] = 'interrupted'
    except subprocess.TimeoutExpired:
        record['status'] = 'process_timeout'
    finally:
        if args.background:
            record['mode'] = ('targeted_mixed_focus' if record['game_foreground_observed'] else
                              'background' if record['focus_observations'] else 'targeted_focus_unobserved')
        record['autonomous_seconds'] = round(time.monotonic()-started, 2) if started is not None else 0
        record['verified_loot_cycles'] = sum(c['outcome'] == 'loot_collected' for c in record['cycles'])
        record['unverified_retrievals'] = sum(c['outcome'] == 'retrieval_unverified' for c in record['cycles'])
        record['perfect_run'] = (record.get('status') == 'completed' and record['autonomous_seconds'] >= 300
                                 and bool(record['cycles']) and record['verified_loot_cycles'] == len(record['cycles']))
        record['finished_at_utc'] = datetime.now(timezone.utc).isoformat()
        (folder/'summary.json').write_text(json.dumps(record, indent=2)+'\n')
        print(json.dumps({k: record[k] for k in ['status','mode','pre_go_runs','autonomous_seconds','verified_loot_cycles','unverified_retrievals','provider_calls','perfect_run']}), flush=True)


if __name__ == '__main__':
    main()

File: experiments/001_wow_fishing/run_test_a.swift
// Compare bite policies with identical preparation, observations and input checks.
//
//   run-test-a [--background] [--jev] [--seconds 60..1800]
//
// Pre-go runs the live helper (/tmp/jev-fishing-live, from build.sh) once with --check; the autonomous
// interval then repeats --execute --prepared. Test B (--jev) is provider-backed bite decisions with the
// same deterministic preparation. Built by build.sh with dotenv.swift and ../002_wow_visual/runtime/JSON.swift.
import CryptoKit
import Foundation

let runnerRoot = URL(fileURLWithPath: #filePath).deletingLastPathComponent().deletingLastPathComponent().deletingLastPathComponent()
let liveBinary = "/tmp/jev-fishing-live"
let hashedSources = ["live.swift", "decision.swift", "self_tests.swift", "motion.swift", "motion-fixtures.png", "jev.swift", "loot.swift",
                     "loot-close.png", "loot-layout.png", "build.sh", "page-two.png", "rod-icon.png", "split-bobber.png",
                     "split-bobber-before.png", "bobber-no-red.png", "bobber-no-red-before.png", "run_test_a.swift",
                     "probes/background-click/Adapter.swift", "probes/background-click/NativeWindowServerPreparation.swift",
                     "probes/background-click/NativeBackgroundClickTransport.swift"]
// A held bite that never returns is still an unconfirmed target before any click.
let retryable: Set<String> = ["stopped_no_visible_float", "stopped_target_lost", "stopped_bite_not_recovered"]
let terminalEvents: Set<String> = ["loot_item_unconfirmed", "loot_not_cleared", "loot_layout_unconfirmed", "loot_batch_limit",
                                   "retrieval_unverified"]

struct Invocation {
    var code: Int32 = 0
    var log = ""
    var events: [JSON] = []
}

struct RunnerUsage: Error, CustomStringConvertible { let description: String }
struct RunnerTimeout: Error {}
struct RunnerInterrupted: Error {}

/// Everything the runner does to the world, so the checks can replace it.
struct RunnerHost {
    var invoke: (_ arguments: [String], _ timeout: Double, _ environment: [String: String]) throws -> Invocation
    var now: () -> Double
    var sleep: (Double) -> Void
    var say: (String) -> Void
    var interrupted: () -> Bool
}

func utcNow() -> String {
    let format = DateFormatter()
    format.dateFormat = "yyyy-MM-dd'T'HH:mm:ss.SSSSSSxxxxx"
    format.timeZone = TimeZone(identifier: "UTC")
    format.locale = Locale(identifier: "en_US_POSIX")
    return format.string(from: Date())
}

func rounded2(_ x: Double) -> JSON { .double(JSON.round(x, 2)) }

func event(_ e: JSON) -> String { e["event"]?.string ?? "" }

/// Run Test A (rules) or B (Jev) and return the summary it wrote. A timeout or an interruption ends the
/// run with that status; any other error still writes the summary, then propagates.
@discardableResult
func runTestA(_ arguments: [String], root: URL, environment: [String: String], host: RunnerHost) throws -> JSON {
    var background = false, jev = false, seconds = 300
    var rest = arguments[...]
    while let flag = rest.popFirst() {
        switch flag {
        case "--background": background = true
        case "--jev": jev = true
        case "--seconds":
            guard let value = rest.popFirst().flatMap(Int.init) else { throw RunnerUsage(description: "--seconds needs a whole number") }
            seconds = value
        default: throw RunnerUsage(description: "unknown argument " + flag)
        }
    }
    // A live-input duration, so reject a typo rather than dispatch input for hours.
    guard (60...1800).contains(seconds) else { throw RunnerUsage(description: "--seconds must be between 60 and 1800") }
    var env = environment
    if jev {
        guard let key = typesafeKey(root: root, environment: environment) else {
            throw RunnerUsage(description: "Set TYPESAFE_API_KEY locally before Test B")
        }
        env["TYPESAFE_API_KEY"] = key
        env["JEV_CALL_LIMIT"] = "120"
    }
    let label = jev ? "B" : "A"
    let mode = background ? ["--background"] : []
    let policyMode = mode + (jev ? ["--jev"] : [])
    let clock = DateFormatter()
    clock.dateFormat = "yyyyMMdd'T'HHmmss'Z'"
    clock.timeZone = TimeZone(identifier: "UTC")
    clock.locale = Locale(identifier: "en_US_POSIX")
    let id = UUID().uuidString.lowercased().replacingOccurrences(of: "-", with: "").prefix(6)
    let folder = root.appendingPathComponent("runs/001_wow_fishing/test_\(label.lowercased())_\(clock.string(from: Date()))_\(id)")
    try FileManager.default.createDirectory(at: folder, withIntermediateDirectories: true)
    var record = JSON.object([
        ("protocol", .string("anchored-pixel-bite-only-v1")), ("comparison_scope", .string("bite_policy_only")),
        ("policy", .string(jev ? "jev" : "rules")), ("call_budget", .int(jev ? 120 : 0)), ("status", .string("running")),
        ("mode", .string(background ? "targeted" : "foreground")),
        ("input_mode", .string(background ? "targeted_without_activation" : "foreground_without_activation")),
        ("game_foreground_observed", .bool(false)), ("focus_observations", .int(0)), ("target_seconds", .int(seconds)),
        ("pre_go_runs", .int(1)), ("provider_calls", .int(0)), ("started_at_utc", .string(utcNow())), ("cycles", .array([]))])
    let source = root.appendingPathComponent("experiments/001_wow_fishing")
    record = record.setting("source_hashes", .object(try hashedSources.map { name in
        (name, .string(SHA256.hash(data: try Data(contentsOf: source.appendingPathComponent(name))).map { String(format: "%02x", $0) }.joined()))
    }))
    host.say("Test \(label) \(record["mode"]!.string!); results: \(folder.path)")

    func int(_ key: String) -> Int { if case let .int(i) = record[key] { return i }; return 0 }
    func account(_ events: [JSON]) {
        record = record.setting("provider_calls", .int(int("provider_calls") + events.filter { event($0) == "jev_request" }.count))
        for e in events {
            if event(e) == "focus_observed" {
                record = record.setting("focus_observations", .int(int("focus_observations") + 1))
                record = record.setting("game_foreground_observed",
                                        .bool(record["game_foreground_observed"] == .bool(true) || e["game_foreground"] == .bool(true)))
            }
            if event(e) == "jev_response" {
                record = record.setting("request_seconds", .array((record["request_seconds"]?.items ?? []) + [e["request_seconds"] ?? .null]))
                for (name, count) in e["response"]?["usage"]?.pairs ?? [] {
                    switch (record[name], count) {
                    case let (.int(a)?, .int(b)): record = record.setting(name, .int(a + b))
                    case (nil, _): record = record.setting(name, count)
                    default: record = record.setting(name, .double((record[name]?.number ?? 0) + (count.number ?? 0)))
                    }
                }
            }
        }
        if jev { env["JEV_CALL_LIMIT"] = String(max(0, 120 - int("provider_calls"))) }
    }
    func call(_ arguments: [String], _ timeout: Double) throws -> Invocation {
        let result = try host.invoke(arguments, timeout, env)
        if host.interrupted() { throw RunnerInterrupted() }
        return result
    }
    var cycles: [JSON] = []
    var started: Double?
    var failure: Error?
    do {
        let prego = try call(["--check"] + mode, 120)
        account(prego.events)
        try prego.log.write(to: folder.appendingPathComponent("pre-go.log"), atomically: true, encoding: .utf8)
        if prego.code != 0 || !prego.events.contains(where: { event($0) == "pre_go_pass" }) {
            record = record.setting("status", .string("pre_go_failed"))
        } else {
            started = host.now()
            host.say("Pre-go passed once; autonomous \(seconds)-second test started.")
            var failures = 0, stopped = false
            // Finish an in-flight cast rather than kill it with a mouse button held. This bounds normal
            // completion to the interval plus at most one 45-second cycle.
            while host.now() - started! < Double(seconds) {
                let number = cycles.count + 1
                let result = try call(["--execute", "--prepared"] + policyMode, 45)
                account(result.events)
                try result.log.write(to: folder.appendingPathComponent(String(format: "cycle-%02d.log", number)), atomically: true, encoding: .utf8)
                let events = result.events
                let loot = events.first { event($0) == "loot_collected" }
                let native = loot != nil ? "loot_collected" : events.last.map(event) ?? "process_failed"
                let unconfirmed = result.code == 0 && retryable.contains(native) && !events.contains { event($0) == "right_click" }
                let cycle = JSON.object([
                    ("number", .int(number)), ("outcome", .string(unconfirmed ? "target_unconfirmed" : native)),
                    ("native_outcome", .string(native)), ("exit_code", .int(Int(result.code))),
                    ("elapsed_seconds", rounded2(host.now() - started!)),
                    ("run_path", events.first { event($0) == "ready" }?["output"] ?? .null),
                    ("item", loot?["item"] ?? .null), ("labels_observed", loot.map { $0["labels_observed"] ?? .array([]) } ?? .array([])),
                    ("loot_clicks", loot.map { $0["loot_clicks"] ?? .int(1) } ?? .int(0))])
                cycles.append(cycle)
                record = record.setting("cycles", .array(cycles))
                host.say(cycle.text())
                failures = loot != nil ? 0 : failures + 1
                if failures >= 3 {
                    record = record.setting("status", .string("stopped_after_three_consecutive_failures"))
                    stopped = true
                    break
                }
                // Only unconfirmed targets before input may retry within the three-failure bound.
                // Camera, geometry, provider and input stops remain terminal.
                if result.code != 0 || events.contains(where: { e in
                    (event(e).hasPrefix("stopped_") && !(unconfirmed && retryable.contains(event(e)))) || terminalEvents.contains(event(e))
                }) {
                    record = record.setting("status", .string("stopped_for_review"))
                    stopped = true
                    break
                }
                host.sleep(0.5)
                if host.interrupted() { throw RunnerInterrupted() }
            }
            if !stopped {
                let caught = cycles.contains { $0["outcome"] == .string("loot_collected") }
                record = record.setting("status", .string(caught ? "completed" : "no_verified_catches"))
            }
        }
    } catch is RunnerInterrupted {
        record = record.setting("status", .string("interrupted"))
    } catch is RunnerTimeout {
        record = record.setting("status", .string("process_timeout"))
    } catch {
        failure = error
    }
    if background {
        record = record.setting("mode", .string(record["game_foreground_observed"] == .bool(true) ? "targeted_mixed_focus"
                                                : int("focus_observations") > 0 ? "background" : "targeted_focus_unobserved"))
    }
    let autonomous = started.map { rounded2(host.now() - $0) } ?? .int(0)
    let verified = cycles.filter { $0["outcome"] == .string("loot_collected") }.count
    record = record.setting("autonomous_seconds", autonomous)
        .setting("verified_loot_cycles", .int(verified))
        .setting("unverified_retrievals", .int(cycles.filter { $0["outcome"] == .string("retrieval_unverified") }.count))
    record = record.setting("perfect_run", .bool(record["status"] == .string("completed") && (autonomous.number ?? 0) >= Double(seconds)
                                                 && !cycles.isEmpty && verified == cycles.count))
        .setting("finished_at_utc", .string(utcNow()))
    try (record.text(indent: 2) + "\n").write(to: folder.appendingPathComponent("summary.json"), atomically: true, encoding: .utf8)
    host.say(JSON.object(["status", "mode", "pre_go_runs", "autonomous_seconds", "verified_loot_cycles", "unverified_retrievals",
                          "provider_calls", "perfect_run"].map { ($0, record[$0]!) }).text())
    if let failure { throw failure }
    return record
}

// Set by SIGINT. A handler, not SIG_IGN: an ignored signal would stay ignored in the live helper, which must
// get the owner's Ctrl-C itself to release its input.
nonisolated(unsafe) var sigint: sig_atomic_t = 0

/// The live helper, its output in files (a pipe could fill and stall it), bounded by `timeout`.
func invokeLive(_ arguments: [String], _ timeout: Double, _ environment: [String: String]) throws -> Invocation {
    let scratch = FileManager.default.temporaryDirectory.appendingPathComponent("run-test-a-\(UUID().uuidString)")
    try FileManager.default.createDirectory(at: scratch, withIntermediateDirectories: true)
    defer { try? FileManager.default.removeItem(at: scratch) }
    let outURL = scratch.appendingPathComponent("out"), errURL = scratch.appendingPathComponent("err")
    FileManager.default.createFile(atPath: outURL.path, contents: nil)
    FileManager.default.createFile(atPath: errURL.path, contents: nil)
    let process = Process()
    process.executableURL = URL(fileURLWithPath: liveBinary)
    process.arguments = arguments
    process.currentDirectoryURL = runnerRoot
    process.environment = environment
    process.standardOutput = try FileHandle(forWritingTo: outURL)
    process.standardError = try FileHandle(forWritingTo: errURL)
    try process.run()
    let deadline = ProcessInfo.processInfo.systemUptime + timeout
    while process.isRunning {
        if ProcessInfo.processInfo.systemUptime > deadline {
            process.terminate()
            let grace = ProcessInfo.processInfo.systemUptime + 5
            while process.isRunning && ProcessInfo.processInfo.systemUptime < grace { usleep(50_000) }
            if process.isRunning { kill(process.processIdentifier, SIGKILL) }
            process.waitUntilExit()
            throw RunnerTimeout()
        }
        usleep(20_000)
    }
    let out = (try? String(contentsOf: outURL, encoding: .utf8)) ?? "", err = (try? String(contentsOf: errURL, encoding: .utf8)) ?? ""
    let events = out.split(whereSeparator: \.isNewline).compactMap { try? JSON.parse(String($0)) }.filter { $0["event"] != nil }
    return Invocation(code: process.terminationStatus, log: out + err, events: events)
}

#if !RUNNER_TESTS
@main
struct RunTestA {
    static func main() {
        signal(SIGINT) { _ in sigint = 1 }
        let host = RunnerHost(invoke: invokeLive, now: { ProcessInfo.processInfo.systemUptime }, sleep: { usleep(UInt32($0 * 1_000_000)) },
                              say: { print($0); fflush(stdout) }, interrupted: { sigint != 0 })
        do {
            try runTestA(Array(CommandLine.arguments.dropFirst()), root: runnerRoot, environment: ProcessInfo.processInfo.environment, host: host)
        } catch {
            FileHandle.standardError.write(Data("\(error)\n".utf8))
            exit(1)
        }
    }
}
#endif
